README
#==============================================================
#
#  Name: README
#
#  Rem:  This is a driver for the Acromag IP330 16-channel 
#        Industry Pack (IP) analog input card. 
#
#  Auth: 29-Aug-2007, Sheng Peng         (PENGS)
#  Rev:  dd-mmm-yyyy, Reviewer's Name    (USERNAME)
#--------------------------------------------------------------
#  Mod:
#        dd-mmm-yyyy, First Lastname     (USERNAME):
#          comment
#
#==============================================================
#
EPICS Verions:
--------------
This driver has currently been built and tested at SLAC with EPICS 3.14.8.2.
This driver has implemented EPICS OSI function calls and is OSI complient.

Operating System:
-------------------
This driver has only been tested  under RTEMS. 

How to use this module in your IOC
------------------
1.	Add the following lines in your iocTop/configure/RELEASE:
IP330=$(EPICS_MODULES)/ip330/$(IP330_MODULE_VERSION)

2.	Copy the following files to your iocTop/<iocName>/<iocName>App/Db directory:
ip330App/Db/IP330.substitutions

3.	Add the following lines in your iocTop/<iocName>/<iocName>App/Db/Makefile:
DB += IP330.db

4.	Add the following lines in your iocTop/<iocName>/<iocName>App/src/Makefile:
<iocName>_DBD += IP330.dbd
<iocName>_LIBS += IP330

5.	Add the following lines in your iocTop/<iocName>/<iocName>App/src/<iocName>Main.cpp:
//  Include these lines to force linker to include xy9660 Ipac carrier
//	Note: There are other ipac carrier modules besides the xy9660.
//	See the ipac module for options.
#include "drvXy9660.h"
extern "C"
int addCarrier( const char * cardParams )
{
    return ipacAddXVME9660( cardParams );
}

6.	Add the following lines in your iocTop/<iocName>/iocBoot/ioc<iocName>/st.cmd:
# Initialize IPAC carrier
# The 2nd parameter is the base I/O address set by the jumpers on your carrier board
# Make sure this address doesn't conflict with other boards.
addCarrier( "0x0000" )

#	Initialize IP330 analog input module
#	See the Acromag IP330 User's manual in ip330/documentation
#	for details on scan modes, triggers, timer, and other options.
#	ip330Create(
#		char	*	cardname,	Unique Identifier "ip330-1"
#		UINT16		carrier,	Ipac Driver carrier card number, 0 for first ipac
#		UINT16		slot,		Slot number on IP carrier, 0..3 for slots A..D
#		char	*	a2d_range,	One of:	"-5to5D" "-10to10D" "0to5D" "0to10D"
#										"-5to5S" "-10to10S" "0to5S" "0to10S"
#		char	*	channels,	Format is "ch%d-ch%d", Ex: "ch0-ch15" or "ch21-ch31"
#								Available range is 0 to 15/31 depending on Differential or
#								Single inputs.
#		UINT32		gainL,		The lower two bits is gain for channel0, then channel 1 ... 15
#		UINT32		gainH,		The lower two bits is gain for channel16, then channel 17 ... 31
#		char	*	scanmode,	Format is "%s-%s-%s", <mode>-<trigger>-Avg<avgCount>
#								mode can be "uniformCont", "uniformSingle", "burstCont",
#									"burstSingle", or "cvtOnExt"
#								trigger can be "Input" or "Output"
#									If mode is "cvtOnExt", trigger must be "Input"
#								avgCount range is 1 to 256 and specifies how many samples to
#									average over
#								For low frequency applications without exacting timing
#									requirements, a scanmode of "burstCont-Output-Avg1" is typical.
#		char	*	timer,		Format is "x*y@8MHz" where x must be [64,255], y must be [1,65535]
#								The timing interval can be calculated from (x*y)/8 in microseconds.
#		UINT8		vector )	Interrupt vector must be unique.  Suggest starting from 0x66
ip330Create( "ai0", 0, 2, "-10to10D", "ch0-ch8", 0, 0, "burstCont-Output-Avg1", "80*3@8MHz", 0x66 )
ip330StartConvertByName( "ai0" )

# Load the EPICS records
dbLoadRecords("db/IP330.db","CARD=ai0")

############################################################

EPICS PV Values:
----------------------
The default names assigned by the db template and substitution files
in this module are shown below.  These can be customized after you copy the
template and substitution files to your ioc.
------------------
PV Name:	Type:		Description:
ai0:in0		bi			Analog Input 0
ai0:in1		bi			Analog Input 1
ai0:in2		bi			Analog Input 2
ai0:in3		bi			Analog Input 3
ai0:in4		bi			Analog Input 4
ai0:in5		bi			Analog Input 5
ai0:in6		bi			Analog Input 6
ai0:in7		bi			Analog Input 7
ai0:in8		bi			Analog Input 8
ai0:in9		bi			Analog Input 9
ai0:in10	bi			Analog Input 10
ai0:in11	bi			Analog Input 11
ai0:in12	bi			Analog Input 12
ai0:in13	bi			Analog Input 13
ai0:in14	bi			Analog Input 14
ai0:in15	bi			Analog Input 15


RELEASE_NOTES
#==============================================================
#
#  Name: RELEASE_NOTES
#
#  Rem: This files describes release notes for
#       the Acromag ip330 driver and device support.      
#
#  Auth: 29-Aug-2007, Sheng Peng         (PENGS)
#  Rev:  dd-mmm-yyyy, Reviewer's Name    (USERNAME)
#--------------------------------------------------------------
#  Mod:
#        20-Mar-2010, S. Hoobler         (SONYA);
#          release ip330-R2-2-0
#        29-Apr-2009, K. Luchini         (LUCHINI);
#          release ip330-R2-0-0
#        21-Feb-2008, K. Luchini         (LUCHINI);
#          release ip330-R1-0-4
#        08-Oct-2007, K. Luchini         (LUCHINI):
#          release ip330-R1-0-3
#          release ip330-R1-0-2
#        30-Aug-2007, K. Luchini         (LUCHINI):
#          release ip330-R1-0-1
#
#==============================================================
#  
R2.6.4:         Vincent Esposito - Oct 17, 2022
                - Using IPAC R2.15-1.0.2.
                - Compile for rhel7
                - Fix warning messages


R2.6.3:         Marcio Donadio - Nov 27, 2019
                - Using IPAC R2.15-1.0.0.
                - Fixed compiler warning messages.  

R2.6.2:         Kukhee Kim - Oct 19, 2019
                - Using IPAC R2.14-0.2.0 

R2.6.1:         Kukhee Kim - Oct 16, 2019
                - Use type-safe DSET and DRVET
                - build with IPAC R2.14-0.2.0

R2.6.0:		Hugo Slepicka - Nov 08, 2017
		- Adjusted the Header file to be true also for Epics 7.
		- Using IPAC R2.14-0.1.0

R2.5.2:		Marcio Donadio - Nov 07, 2017
                - Modified RELEASE files to set according to new standards

R2.5.1:		Marcio Donadio - Nov 07, 2017
		- Using IPAC R2.11-1.5.1

R2.5.0:		Marcio Donadio - Sep 28, 2017
		- IOC shell function ip330StartConvertByName was created, but
		  not registered

R2.4.0:		Marcio Donadio
		- Changes to make code work with both little and big-endian.
		- EPICS 3.15 and GIT workflow.

ip330-R2-3-0:   J. Mock
                build with ipac R2.11-1.5.0, which builds for all targets
                Added registrar support so module works with non-RTEMS OS

ip330-R2-2-6:   K.Luchini
                build with ipac-R2-11_1-2 which builds for all targets
 
ip330-R2-2-5:   S.Condamoor: autosaveField added toIP 330BSA.template  for alarm limits;
                but removed in MAIN_TRUNK and committed without autosave field again per Ernest.

ip330-R2-2-4: Build against ipac-R2-11_1-0 which builds for all targets

ip330-R2-2-3:         Build with ipac-R2-11-lcls1
             
ip330-R2-2-2:   Build against EPICS R3-14-12 and use ipac-R2-8-lcls8

ip330-R2-2-1:   07-Sep-2011, K. Luchini  (LUCHINI):
                Upgrade ipac from ipac-R2-8-lcls7 to ipac-R2-11-lcls1 

ip330-R2-2-0:   Build with base-R3-14-8-2-lcls6
                Build with rtems-4.9.4

ip330-R2-1-3:   Replace sprintf

ip330-R2-1-2:   Build with base-R3-14-8-2-lcls5
                Build with rtems-4.9.3

ip330-R2-0-1:   Add IP330BSA_TSE.template
		Build with base-R3-14-9-lcls3 (RTEMS 4.9.1)

ip330-R2-0-1:   Change the record to high cb priority
		Build with base-R3-14-9-lcls3 (RTEMS 4.9.1)

ip330-R2-0-0:   Build with base-R3-14-9-lcls3 (RTEMS 4.9.1)

ip330-R1-0-2-1: Small bug fix: Typo for determing the number of single-ended
                channels was fixed. 
                Build against  EPICS R3.14.8.2 and ipac-R2-8-lcls3
                RTEMS version 4.7.1

ip330-R1-0-4:   Build with base-R3-14-9-lcls1 (RTEMS 4.7.1)
                Build with ipac-R2-8-lcls4
                Replace CONFIG_APP with CONFIG_SITE
                Add RELEASE_SITE
                
ip330-R1-0-3:   Build with base-R3-14-9-lcls1 (RTEMS 4.7.1)
                Build with extensions-R3-14-9
                Build with ipac-R2-8-lcls2

ip330-R1-0-2:   Build with base-R3-14-8-2 (RTEMS 4.7.1)
                Build with ipac-R2-8-lcls1

ip330-R1-0-0:   Build with ipac-R1-0-2 instead of ipac-R1-0-1

ip330-R1-0-0:   Initial release


ip330App/src/devAiIP330.c
/****************************************************************/
/* $Id: devAiIP330.c,v 1.5 2006/08/10 16:18:36 pengs Exp $      */
/* This file implements AI record device support for IP330 ADC  */
/* Author: Sheng Peng, pengs@slac.stanford.edu, 650-926-3847    */
/****************************************************************/
#include <stdio.h>
#include <string.h>

#include <epicsVersion.h>

#if ((EPICS_VERSION==3 && EPICS_REVISION>=14) || EPICS_VERSION >3)
#include <epicsExport.h>
#endif

#include "errlog.h"
#include <devLib.h>
#include <dbAccess.h>
#include <dbScan.h>
#include <callback.h>
#include <cvtTable.h>
#include <link.h>
#include <recSup.h>
#include <recGbl.h>
#include <devSup.h>
#include <drvSup.h>
#include <dbCommon.h>
#include <alarm.h>
#include <cantProceed.h>
#include <aiRecord.h>

#include <ptypes.h>
#include <drvIP330Lib.h>

#define MAX_CA_STRING_SIZE (40)

/* define function flags */
typedef enum {
        IP330_AI_DATA,
} IP330FUNC;

static struct PARAM_MAP
{
        char param[MAX_CA_STRING_SIZE];
        int  funcflag;
} param_map[1] = {
    {"DATA", IP330_AI_DATA}
};
#define N_PARAM_MAP (sizeof(param_map)/sizeof(struct PARAM_MAP))

typedef struct IP330_DEVDATA
{
    IP330_ID	pcard;
    UINT16	chnlnum;
    int		funcflag;
} IP330_DEVDATA;

/* This function will be called by all device support */
/* The memory for IP330_DEVDATA will be malloced inside */
static int IP330_DevData_Init(dbCommon * precord, char * ioString)
{
    int		count;
    int		loop;

    char	cardname[MAX_CA_STRING_SIZE];
    IP330_ID	pcard;
    int		chnlnum;
    char	param[MAX_CA_STRING_SIZE];
    int		funcflag = 0;

    IP330_DEVDATA *   pdevdata;

    /* param check */
    if(precord == NULL || ioString == NULL)
    {
        if(!precord) errlogPrintf("No legal record pointer!\n");
        if(!ioString) errlogPrintf("No INP/OUT field for record %s!\n", precord->name);
        return -1;
    }

    /* analyze INP/OUT string */
    count = sscanf(ioString, "%[^:]:%i:%[^:]", cardname, &chnlnum, param);
    if (count != 3)
    {
        errlogPrintf("Record %s INP/OUT string %s format is illegal!\n", precord->name, ioString);
        return -1;
    }

    pcard = ip330GetByName(cardname);
    if( !pcard )
    {
        errlogPrintf("Record %s IP330 %s is not registered!\n", precord->name, cardname);
        return -1;
    }

    if(chnlnum < 0 || chnlnum > 32 )
    {/* chnlnum is UINT16, more accurate check again start/end channel will be done in ip330Read */
        errlogPrintf("Record %s channel number %d is out of range for IP330 %s!\n", precord->name, chnlnum, cardname);
        return -1;
    }

    for(loop=0; loop<N_PARAM_MAP; loop++)
    {
        if( 0 == strcmp(param_map[loop].param, param) )
        {
            funcflag = param_map[loop].funcflag;
            break;
        }
    }
    if(loop >= N_PARAM_MAP)
    {
        errlogPrintf("Record %s param %s is illegal!\n", precord->name, param);
        return -1;
    }

    pdevdata = (IP330_DEVDATA *)callocMustSucceed(1, sizeof(IP330_DEVDATA), "Init record for IP330");

    pdevdata->pcard = pcard;
    pdevdata->chnlnum = chnlnum;
    pdevdata->funcflag = funcflag;

    precord->dpvt = (void *)pdevdata;
    return 0;
}


static long init_ai( struct dbCommon * p)
{
    struct aiRecord *pai = (struct aiRecord *)p;

    pai->dpvt = NULL;

    if (pai->inp.type!=INST_IO)
    {
        recGblRecordError(S_db_badField, (void *)pai, "devAiIP330 Init_record, Illegal INP");
        pai->pact=TRUE;
        return (S_db_badField);
    }

    pai->eslo = (pai->eguf - pai->egul)/(float)0x10000;
    pai->roff = 0x0;

    if(IP330_DevData_Init((dbCommon *) pai, pai->inp.value.instio.string) != 0)
    {
        errlogPrintf("Fail to init devdata for record %s!\n", pai->name);
        recGblRecordError(S_db_badField, (void *) pai, "Init devdata Error");
        pai->pact = TRUE;
        return (S_db_badField);
    }

    return 0;
}


/** for sync scan records  **/
static long ai_ioint_info(int cmd,struct dbCommon *p,IOSCANPVT *iopvt)
{
    struct aiRecord *pai = (struct aiRecord *)p;

    IP330_DEVDATA * pdevdata = (IP330_DEVDATA *)(pai->dpvt);

    *iopvt = *ip330GetIoScanPVT(pdevdata->pcard);
    return 0;
}

static long read_ai(struct aiRecord *pai)
{
    IP330_DEVDATA * pdevdata = (IP330_DEVDATA *)(pai->dpvt);

    int status=-1;
    signed int tmp;

    switch(pdevdata->funcflag)
    {
    case IP330_AI_DATA:
        status = ip330Read(pdevdata->pcard, pdevdata->chnlnum, &tmp);
        break;
    }

    if(status)
    {
        pai->udf=TRUE;
        recGblSetSevr(pai, READ_ALARM, INVALID_ALARM);
        return -1;
    }
    else
    {
        pai->rval = tmp;
        pai->udf=FALSE;
        return 0;/******** convert ****/
    }
}

static long ai_lincvt(struct aiRecord   *pai, int after)
{

        if(!after) return(0);
        /* set linear conversion slope*/
        pai->eslo = (pai->eguf - pai->egul)/(float)0x10000;
        pai->roff = 0x0;
        return(0);
}


#ifndef USE_TYPED_DSET
struct IP330_DEV_SUP_SET
{
    long            number;
    DEVSUPFUN       report;
    DEVSUPFUN       init;
    DEVSUPFUN       init_record;
    DEVSUPFUN       get_ioint_info;
    DEVSUPFUN       read_ai;
    DEVSUPFUN       special_linconv;
} devAiIP330 = {6, NULL, NULL, init_ai, ai_ioint_info, read_ai, ai_lincvt};
#else
struct
{
    dset          commmon;
    long(*read_ai)(struct aiRecord *prec);
    long(*linconv)(struct aiRecord *prec, int after);
} devAiIP330 = { {6, NULL, NULL, init_ai, ai_ioint_info}, read_ai, ai_lincvt};
#endif


#if ((EPICS_VERSION==3 && EPICS_REVISION>=14) || EPICS_VERSION>3)
epicsExportAddress(dset, devAiIP330);
#endif



ip330App/src/drvIP330Lib.h
#ifndef _INCLUDE_DRV_IP330_LIB_H_
#define _INCLUDE_DRV_IP330_LIB_H_

/****************************************************************/
/* $Id: drvIP330Lib.h,v 1.1.1.1 2006/08/10 16:18:36 luchini Exp $     */
/* This file defines the user interface of IP330 support        */
/* Author: Sheng Peng, pengs@slac.stanford.edu, 650-926-3847    */
/****************************************************************/

#ifdef __cplusplus
extern "C" {
#endif  /* __cplusplus */

#include <dbScan.h>
#include <epicsTime.h>

#include "ptypes.h"

typedef struct IP330_CARD * IP330_ID;

int ip330Create (char *cardname, UINT16 carrier, UINT16 slot, char *adcrange, char * channels, UINT32 gainL, UINT32 gainH, char *scanmode, char * timer, UINT8 vector, char *trig, char *delay, char *sync, int bldID);

void ip330Configure(IP330_ID pcard);
void ip330Calibrate(IP330_ID pcard);
void ip330ThreadStart(IP330_ID pcard);

IP330_ID ip330GetByName(char * cardname);
IP330_ID ip330GetByLocation(UINT16 carrier, UINT16 slot);

int ip330Read(IP330_ID pcard, UINT16 channel, signed int * pvalue, epicsTimeStamp *ts);
IOSCANPVT * ip330GetIoScanPVT(IP330_ID pcard);

void ip330StartConvert(IP330_ID pcard);
void ip330StartConvertByName(char * cardname);

#ifdef __cplusplus
}
#endif  /* __cplusplus */


#endif


ip330App/src/ip330sync.cpp
#define _NO_IP330_STATIC_
#include"drvIP330Private.h"
#include"bldPvClient.h"
#include<stdio.h>
#include<string.h>
#include<math.h>

#define IP330_THREAD_PRIORITY (epicsThreadPriorityMedium)
#define IP330_THREAD_STACK	(0x20000)

epicsMutexId    ip330SyncObject::master_lock = NULL;

static int ThreadStart(void *param)
{
    ip330SyncObject *sobj = new ip330SyncObject((IP330_CARD *)param);
    sobj->Wait();         /* Wait for a go! */
    return sobj->poll();  /* This never returns. */
}

extern "C" void ip330ThreadStart(IP330_CARD *ip330)
{
    ip330SyncObject::Init();
    ip330->sync_thread = epicsThreadMustCreate(ip330->cardname, IP330_THREAD_PRIORITY,
                                               IP330_THREAD_STACK,
                                               (EPICSTHREADFUNC)ThreadStart, (void *) ip330);
}

void ip330SyncObject::Init(void)
{
    if (!master_lock) {
        master_lock = epicsMutexMustCreate();
        epicsMutexLock(master_lock);
    }
}

void ip330SyncObject::Wait(void)
{
    epicsMutexLock(master_lock);
    epicsMutexUnlock(master_lock);
}

void ip330SyncObject::StartAll(void)
{
    epicsMutexUnlock(master_lock);
}

ip330SyncObject::ip330SyncObject(IP330_CARD *_ip330)
{
    ip330 = _ip330;
    SetParams(ip330->trig, ip330->gen, ip330->delay, ip330->sync);
}

DataObject *ip330SyncObject::Acquire(void)
{
    while (ip330->rd == ip330->wr) {
        epicsEventWait(ip330->data_avail);
    }
    return new DataObject(ip330->data[ip330->rd++ & N_DATABUF_MASK]);
}

int ip330SyncObject::CheckError(DataObject *dobj)
{
    return 0;
}

#ifdef	ENABLE_CONTEXT_TIMER_DIAGS
static ContextTimerMax ip330ISRLoopContextTimer( "ip330ISRLoop" );
static ContextTimerMax ip330ISRUniformContContextTimer( "ip330ISRUniformCont" );
#endif	//	ENABLE_CONTEXT_TIMER_DIAGS

void ip330SyncObject::QueueData(DataObject *dobj, epicsTimeStamp &evt_time)
{
    int loop, wr = ip330->sum_wr;
    volatile IP330_HW_MAP	*pHardware = ip330->pHardware;	/* controller registers */
    UINT16 *data = (UINT16 *)dobj->data;

    /* Load sum_data, if reach avg_times and cont. mode needs reset, stop scan */
    for(loop = ip330->start_channel; loop <= ip330->end_channel; loop++) {
        UINT32 tmp;
        UINT32 tmp_avgtimes, tmp_sum;

        tmp = ip330->sum_data[wr][loop];
        tmp_sum = (tmp & 0x00FFFFFF);
        tmp_avgtimes = ((tmp & 0xFF000000) >> 24) + 1;

        if( tmp == 0xFFFFFFFF || tmp_avgtimes >= ip330->avg_times) {
            /* No data or already reached the average times */
            tmp_sum = 0;
            tmp_avgtimes = 0;
        }

        if((data[MAX_IP330_CHANNELS] & 1) &&  (tmp_avgtimes+1) <= ip330->avg_times) {
            tmp_sum += data[loop];
            tmp_avgtimes += 1;
        }
        if((data[MAX_IP330_CHANNELS] & 2) &&  (tmp_avgtimes+1) <= ip330->avg_times) {
            tmp_sum += data[loop+16];
            tmp_avgtimes += 1;
        }
        ip330->sum_data[wr][loop] = ( (tmp_sum & 0x00FFFFFF) | ((tmp_avgtimes - 1) << 24) );

        if ( tmp_avgtimes >= ip330->avg_times && loop == ip330->end_channel ) {
            if ((ip330->avg_rst != 0) && (ip330->scan_mode == SCAN_MODE_UNIFORMCONT)) {
#ifdef	ENABLE_CONTEXT_TIMER_DIAGS
                ContextTimer	contextTimer( ip330ISRUniformContContextTimer );
#endif	//	ENABLE_CONTEXT_TIMER_DIAGS
                /* Save the current control register settings */
                UINT16		saved_ctrl = pHardware->controlReg;
                pHardware->controlReg = (saved_ctrl) & 0xF8FF; /* Disable scan */
                while( pHardware->controlReg & 0x0700 ) {	/* How long are we stuck in this loop? */
#ifdef	ENABLE_CONTEXT_TIMER_DIAGS
                    ContextTimer	contextTimer( ip330ISRLoopContextTimer );
#endif	//	ENABLE_CONTEXT_TIMER_DIAGS
                }

                /* Work around hardware bug of uniform continuous mode */
                pHardware->startChannel = ip330->start_channel;
                pHardware->controlReg	= saved_ctrl;
                pHardware->startConvert	= 1;
            }
            /* OK, make this available for the reader! */
            epicsMutexLock(ip330->lock);
            ip330->sum_time = evt_time;
            if (ip330->bldID)
                SendBld(wr);
            ip330->sum_wr = 1 - wr;
            epicsMutexUnlock(ip330->lock);
            scanIoRequest(ip330->ioscan);
        }
    }
}

void ip330SyncObject::SendBld(int wr)
{
    char buf[16*sizeof(double) + sizeof(int)];
    int off = 0;
    int loop, i, cnt;
    double v;

    cnt = ip330->end_channel + 1 - ip330->start_channel;
    memcpy(&buf[off], &cnt, sizeof(int));
    off += sizeof(int);
    for(i = 0, loop = ip330->start_channel; loop <= ip330->end_channel; i++, loop++) {
        UINT32	tmp = ip330->sum_data[wr][loop];

        if(tmp == 0xFFFFFFFF) {
            v = NAN;
            memcpy(&buf[off], &v, sizeof(double));
            off += sizeof(double);
            continue;
        }

        double	tmp_sum		 = tmp & 0x00FFFFFF;
	double	tmp_avgtimes = ( (tmp & 0xFF000000) >> 24 ) + 1;
	double	value		 = ip330->adj_slope[ip330->gain[loop]]
                                    * ( tmp_sum/tmp_avgtimes + ip330->adj_offset[ip330->gain[loop]] );
        int pvalue = static_cast<int>( value );
        v = pvalue * ip330->eslo + ip330->eoff;
        memcpy(&buf[off], &v, sizeof(double));
        off += sizeof(double);
    }
    BldSendPacket(ip330->bldClient, ip330->bldID, 0x10000 | 94, &ip330->sum_time, &buf,
                  sizeof(unsigned int) + sizeof(double) * cnt);
}

void ip330SyncObject::DebugPrint(DataObject *dobj)
{
}


ip330App/src/drvIP330Private.h
#ifndef _INCLUDE_DRV_IP330_PRIVATE_H_
#define _INCLUDE_DRV_IP330_PRIVATE_H_

#ifdef __cplusplus
extern "C" {
#endif  /* __cplusplus */

/******************************************************************/
/* $Id: drvIP330Private.h,v 1.1.1.1 2006/08/10 16:18:05 luchini Exp $   */
/* This file defines the internal hw/sw struct of IP330 ADC module*/
/* Author: Sheng Peng, pengs@slac.stanford.edu, 650-926-3847      */
/******************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#include <epicsVersion.h>
#if (EPICS_VERSION>=3 && EPICS_REVISION>=14) || EPICS_VERSION>=7

#include <epicsMutex.h>
#include <epicsThread.h>
#include <epicsString.h>
#include <epicsInterrupt.h>
#include <cantProceed.h>
#include <epicsExport.h>
#include <drvSup.h>
#include <dbScan.h>
#include <ellLib.h>

#include "drvIpac.h"

#include "ptypes.h"
#include "IP330Constant.h"

#else
#error "You need EPICS 3.14 or above because we need OSI support!"
#endif

#define IP330_DRV_VERSION "IP330 driver V1.0"

#define DEBUG_MSG_SIZE 256
/* *******************************************************************/
/* Hardware registers for ADC channels */
/* The IPAC drivers were written for PowerPC which is big-endian. */
/* Intel uses little-endian. This has an effect on how the */
/* IPAC driver sees the IPAC in memory. */
/* To resolve this use compiler */
/* pre-defined macros such as "__BIG_ENDIAN__" below */
/* ******************************************************************/
typedef struct IP330_HW_MAP
{
  volatile UINT16   controlReg;         /* board control register */

#ifdef  __BIG_ENDIAN__
  volatile UINT8    timerPrescaler;     /* timer prescaler register */
  volatile UINT8    intVector;          /* interrupt vector register */
#else
  volatile UINT8    intVector;          /* interrupt vector register */
  volatile UINT8    timerPrescaler;     /* timer prescaler register */
#endif
  volatile UINT16   conversionTimer;    /* conversion timer count register */
#ifdef  __BIG_ENDIAN__
  volatile UINT8    endChannel;         /* end channel register */
  volatile UINT8    startChannel;       /* start channel register */
#else
  volatile UINT8    startChannel;       /* start channel register */
  volatile UINT8    endChannel;         /* end channel register */
#endif
  volatile UINT16   newData[2];         /* new data register */
  volatile UINT16   missedData[2];      /* missed data register */
  volatile UINT16   startConvert;       /* start conversion register */
  volatile UINT16   notUsed1[7];        /* not used */
  volatile UINT8    gain[32];           /* gain array */
  volatile UINT16   data[32];           /* data area */
} IP330_HW_MAP;
/* We don't use __attribute__ ((packed)) here because it is already naturally aligned */
/* And packed with -mstrict-align will make access to byte-access, it will hurt */


/* device driver ID structure */

typedef ELLLIST IP330_CARD_LIST;

typedef struct IP330_CARD
{
    ELLNODE                     node;		/* Link List Node */

    epicsMutexId                lock;
    double                      adj_slope[N_GAINS];
    double                      adj_offset[N_GAINS];

    UINT32                      sum_data[MAX_IP330_CHANNELS]; /* The highest byte +1 is number of samples, then lower 24 bits holds summary of up to 255 samples */
                                                              /* The maximum will be 0xFFFFFF00, so 0xFFFFFFFF is used to indicate no data */
    IOSCANPVT                   ioscan;         /* Trigger EPICS record */

    char                        * cardname;	/* Card identification */
    UINT16                      carrier;	/* Industry Pack Carrier Index */
    UINT16                      slot;		/* Slot number on carrier */

    UINT16                      inp_range;	/* Input range */
    UINT16                      inp_typ;	/* Input type, diff or single end */
    UINT32                      num_chnl;	/* Number of Channels, 16 for differential mode, 32 for single-end mode */

    UINT8                       start_channel;	/* start channel, from 0 */
    UINT8                       end_channel;	/* end channel, up to 15 or 31 */
    UINT32                      chnl_mask;	/* mask to screen missedData and newData flag */

    UINT8                       gain[MAX_IP330_CHANNELS];

    UINT16                      scan_mode;	/* Scan Mode */
    UINT16                      trg_dir;	/* Trigger Direction */
    UINT32                      avg_times;	/* Average times */
    UINT32                      avg_rst;	/* Reset after Average */

    UINT8                       timer_prescaler;/* Tomer Prescaler */
    UINT16                      conversion_timer;/* Conversion Timer */

    UINT8                       vector;		/* Interrupt vector */

    volatile IP330_HW_MAP       *pHardware;	/* controller registers */

    char                        debug_msg[DEBUG_MSG_SIZE];
} IP330_CARD;


#ifdef __cplusplus
}
#endif  /* __cplusplus */

#endif


ip330App/src/IP330Constant.h
#ifndef _INCLUDE_IP330_CONSTANT_H_
#define _INCLUDE_IP330_CONSTANT_H_

#ifdef __cplusplus
extern "C" {
#endif  /* __cplusplus */

/****************************************************************/
/* $Id: IP330Constant.h,v 1.2 2006/12/09 00:14:11 pengs Exp $   */
/* This file defines the constant of IP330 ADC module           */
/* Don't modify definition below this line                      */
/* Author: Sheng Peng, pengs@slac.stanford.edu, 650-926-3847    */
/****************************************************************/


#include "ptypes.h"

/* Manufacturers assigned values */
#define IP_MANUFACTURER_ACROMAG		0xa3
#define IP_MODEL_ADC_330		0x11

/* Define Control Register Bits */
#define CTRL_REG_STRGHT_BINARY		0x0002	/* We will always use Straight Binary format */

#define CTRL_REG_TRGDIR_SHFT		2

#define CTRL_REG_INPTYP_SHFT		3

#define CTRL_REG_SCANMODE_SHFT		8

#define CTRL_REG_TIMR_ENBL		0x0800	/* We will always enable timer if not cvtOnExt or burstSingle mode */

#define CTRL_REG_INTR_CTRL		0x2000	/* We always use interrupt after all channel conversion */
/* Define Control Register Bits Done */


#define MAX_IP330_CHANNELS		32

#define N_GAINS				4
#ifndef _NO_IP330_STATIC_
static const double pgaGain[N_GAINS] = {1.0, 2.0, 4.0, 8.0};
#endif

#define N_RANGES			4
#define ADC_RANGE_B5V			0
#define ADC_RANGE_B10V			1
#define ADC_RANGE_U5V			2
#define ADC_RANGE_U10V			3

#define N_INPTYPS			2
#define INP_TYP_DIFF			0
#define INP_TYP_SE			1
#ifndef _NO_IP330_STATIC_
static const char * rangeName[N_RANGES*N_INPTYPS] = {"-5to5D","-10to10D","0to5D","0to10D","-5to5S","-10to10S","0to5S","0to10S"};

static const char * channelsFormat = "ch%d-ch%d";
#endif

#define N_SCANMODES			6
#define SCAN_MODE_DISABLE		0
#define SCAN_MODE_UNIFORMCONT		1
#define SCAN_MODE_UNIFORMSINGLE		2
#define SCAN_MODE_BURSTCONT		3
#define SCAN_MODE_BURSTSINGLE		4
#define SCAN_MODE_CVTONEXT		5
#ifndef _NO_IP330_STATIC_
static const char * scanModeName[N_SCANMODES] = {"disable","uniformCont","uniformSingle","burstCont","burstSingle","cvtOnExt"};
#endif

#define N_TRGDIRS			2
#define TRG_DIR_INPUT			0
#define TRG_DIR_OUTPUT			1
#ifndef _NO_IP330_STATIC_
static const char * trgDirName[N_TRGDIRS] = {"Input","Output"};
#endif

#define MAX_AVG_TIMES			256
#ifndef _NO_IP330_STATIC_
static const char * avgFormat = "Avg%d";
#endif

#define MIN_TIMER_PRESCALER		0x40
#define MAX_TIMER_PRESCALER		0xFF
#define MIN_CONVERSION_TIMER		0x1
#define MAX_CONVERSION_TIMER		0xFFFF
#ifndef _NO_IP330_STATIC_
static const char * timerFormat = "%d*%d@8MHz";

static struct CAL_SETTING {
    double volt_callo;
    double volt_calhi;
    UINT16 ctl_callo;
    UINT16 ctl_calhi;
    double ideal_span;
    double ideal_zero;
} calSettings[N_RANGES][N_GAINS] =
{
    {   {0.0000, 4.9000,  0x043A,  0x041A, 10.0,  -5.0},
        {0.0000, 2.4500,  0x043A,  0x0422, 10.0,  -5.0},
        {0.0000, 1.2250,  0x043A,  0x042A, 10.0,  -5.0},
        {0.0000, 0.6125,  0x043A,  0x0432, 10.0,  -5.0} },

    {   {0.0000, 4.9000,  0x043A,  0x041A, 20.0, -10.0},
        {0.0000, 4.9000,  0x043A,  0x041A, 20.0, -10.0},
        {0.0000, 2.4500,  0x043A,  0x0422, 20.0, -10.0},
        {0.0000, 1.2250,  0x043A,  0x042A, 20.0, -10.0} },

    {   {0.6125, 4.9000,  0x0432,  0x041A,  5.0,   0.0},
        {0.6125, 2.4500,  0x0432,  0x0422,  5.0,   0.0},
        {0.6125, 1.2250,  0x0432,  0x042A,  5.0,   0.0},
        {0.0000, 0.6125,  0x043A,  0x0432,  5.0,   0.0} },

    {   {0.6125, 4.9000,  0x0432,  0x041A, 10.0,   0.0},
        {0.6125, 4.9000,  0x0432,  0x041A, 10.0,   0.0},
        {0.6125, 2.4500,  0x0432,  0x0422, 10.0,   0.0},
        {0.6125, 1.2250,  0x0432,  0x042A, 10.0,   0.0} }
};
#endif

#ifdef __cplusplus
}
#endif  /* __cplusplus */

#endif
